### PR TITLE
Add auto-generated cuDNN depthwise conv heuristic

### PR DESCRIPTION
**Pull request by gderossi**
This PR revisits the depthwise convolution heuristic introduced in #58749, to automate the heuristic generation process and to improve behavior with newer devices and cuDNN versions. The updated heuristic performs about 5-8% better on average than the existing heuristic, and the entire decision tree can be replicated from available benchmarking data.

### AutoHeuristic scripts
I modified the benchmarking script used in #58749 and #22302 to mirror the design of the existing `BenchmarkRunner` for inductor autoheuristics. While the benchmarking and heuristic generator scripts look similar to the existing equivalents for inductor, they don't actually inherit from any of the existing base classes due to substantial differences in data gathering and code generation (ranking kernels vs T/F, Python vs C++, etc). There is definitely a path forward for aligning these and generalizing the `torchgen/_autoheuristic` machinery if that is of interest, but I decided after tinkering with full integration into the existing classes that independent scripts would be much easier to review, for now.

### `torch.backends.cudnn.force` flag
I ran my initial benchmarking on a source build with a flag to force cuDNN on or off for depthwise convolution (from #139699), and this PR includes a modified version of that flag to allow the benchmarking script to work with the source code it's being shipped with while also maintaining prior heuristic behavior.

### Other notes
- Profiling autogen heuristics vs the existing heuristic suggested a slowdown of 3-10%, increasing average runtime on the test CPU from 80 ns to 82-89 ns. Given the time scale of this function relative to the actual convolutions, which are measured in us/ms, I don't think the increased cost is an issue here.
- EDIT: Thanks Eddie for suggesting a CUDAHook approach for SM version- SM version is now used as a decision variable again. ~The current version of the training script averages normalized speedup data from benchmarking runs on different devices. I also tested a version of heuristic autogen which used the SM version of the GPU as a feature in the decision tree, and this yielded a deeper decision tree (\~2700 LOC) with better performance, about 13.9% average speedup on A100. Unfortunately, I was unable to find a simple approach that both satisfied the `ATEN_CPU_GPU_AGNOSTIC` linter and was actually able to build, since I depended on `at::cuda::getCurrentDeviceProperties()` inside `Convolution.cpp`. If there is a viable path forward for this, it would unlock an additional 3-5% average speedup.~

cc @csarofeen @ptrblck @xwang233 @eqy @nWEIdia @msaroufim @jerryzh168 @tinglvv @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01, @ptrblck

### CODE REVIEW

File: aten/src/ATen/autoheuristic/DepthwiseConvHeuristic.h [hunk 1]
```diff
@@ -2,11 +2,15 @@
...
 
 #include <ATen/core/Tensor.h>
+#include <ATen/detail/CUDAHooksInterface.h>
 
 namespace at::native {
 
 template <typename T>
 static bool check_cudnn_depthwise_workload_with_filter(const at::Tensor& input, T stride, const at::Tensor& weight) {
+  static int sm = at::detail::getCUDAHooks().getDeviceCapability();
+  if (sm == 0) return false; // CUDA not available
```
Comment (eqy) [inline aten/src/ATen/autoheuristic/DepthwiseConvHeuristic.h]: we could in theory `TORCH_INTERNAL_ASSERT` this right, as we should not get this far if CUDA was not even available

File: torchgen/_autoheuristic/depthwise_conv/train_decision_depthwiseconv.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
+import argparse
+
+import numpy as np
+import pandas as pd
+from sklearn.metrics import accuracy_score
+from sklearn.tree import DecisionTreeClassifier
```
Comment (malfet) [inline torchgen/_autoheuristic/depthwise_conv/train_decision_depthwiseconv.py:6]: Is there a respective PyTorch functions?

Comment (gderossi) [reply torchgen/_autoheuristic/depthwise_conv/train_decision_depthwiseconv.py:6]: IIUC there is no native decision tree implementation in PyTorch. I could certainly write one but opted just to use sklearn because the existing autoheuristic work does too. The `accuracy_score` function can definitely be done in a couple lines of code without the sklearn dependency. Happy to refactor this without sklearn if it's desirable limit keep these scripts to existing dependencies/PyTorch itself.

Comment (albanD) [reply torchgen/_autoheuristic/depthwise_conv/train_decision_depthwiseconv.py:6]: I am confused here. We should either:
- Check in the output as code in the .h file. But move this logic to the scripts/ folder. We don't want this in torchgen and we don't want to take a dependency on sklearn there.
- Have this generation run as part of the codegen, and it's ok to add the dependency. But then we need to make sure that generation is fast and we should remove the checked in .h file

Comment (gderossi) [reply torchgen/_autoheuristic/depthwise_conv/train_decision_depthwiseconv.py:6]: I think the first option would be preferable here, as this heuristic probably only needs to be regenerated with new cuDNN releases that touch convolution logic or to update hardware support. From what I understand of the existing code for inductor heuristics though, nothing in `_autoheuristic` gets run during codegen- it's done manually (and the artifacts from training are checked in under `torch/_inductor/autoheuristic/artifacts`). Is that correct? And if so, would it make sense to move all `_autoheuristic` into the scripts/ folder (perhaps as a separate PR for clarity)?

**Review by malfet — COMMENTED** [review]
Let's move `backends.cudnn.force` into a separate PR and give it a better name

Also, it clearly needs rebase to get cleaner signal

File: torchgen/_autoheuristic/depthwise_conv/train_decision_depthwiseconv.py (hunk 2 above)
Comment (eellison) [reply torchgen/_autoheuristic/depthwise_conv/train_decision_depthwiseconv.py:6]: I think it would be great to get this landed, and extend the set of autoheuristic capabilities. cc @v0i0 @AmesingFlank. 

i think the comments are resolvable, but it's fair to say we want to figure out exactly what the build process should be. We have precedent for other, artifacts that are generated once and checked in.

**Comment by gderossi** [comment]
Hi @albanD @drisspg, just bumping this for another look now that #176500 has landed. Thanks!

**Comment by gderossi** [comment]
@drisspg rerequesting review

File: torchgen/_autoheuristic/depthwise_conv/gen_data_depthwiseconv.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        for _ in range(self.nb_iters):
+            output = conv(x)
+            x.grad = None
+            conv.weight.grad = None
+            output.backward(g0)
+        self.end_event.record()
+        torch.cuda.synchronize()
+        all_time = self.start_event.elapsed_time(self.end_event) / self.nb_iters
+        bwd_time = all_time - fwd_time
+        return fwd_time, bwd_time, all_time
```
Comment (drisspg) [inline torchgen/_autoheuristic/depthwise_conv/gen_data_depthwiseconv.py:126]: can we reuse some of the benchmark funcs from indcutor which do l2 cache cleraing and other best practices

File: torchgen/_autoheuristic/depthwise_conv/get_depthwiseconv_dataset.sh [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+base_url='https://github.com/gderossi/aten-autoheuristic-datasets/raw/main/'  # @lint-ignore
```
Comment (drisspg) [inline torchgen/_autoheuristic/depthwise_conv/get_depthwiseconv_dataset.sh:3]: hmmmm
can we check this into the repo maybe?  or I guess do we even need this? you can maybe jsut add to the readme. Saying in there liek this was ran on x hardware at this date and produced this data with this sha

File: torchgen/_autoheuristic/depthwise_conv/train_decision_depthwiseconv.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+
+
+class TrainDecisionTreeDepthwiseConv:
+    def __init__(self):
+        self.features = ["sm", "bs", "ch", "w", "filter", "stride"]
+        self.target = "cudnn_speedup_all"
+        self.classes = ["false", "true"]
+        self.output_file = (
+            "../../../aten/src/ATen/autoheuristic/DepthwiseConvHeuristic.h"
```
Comment (drisspg) [inline torchgen/_autoheuristic/depthwise_conv/train_decision_depthwiseconv.py:15]: gotta be some better path shenangins we can do here

File: aten/src/ATen/autoheuristic/DepthwiseConvHeuristic.h [hunk 6]
```diff
@@ -0,0 +1,2727 @@
...
+#pragma once
+
+#include <ATen/core/Tensor.h>
+#include <ATen/detail/CUDAHooksInterface.h>
+
+namespace at::native {
+
+template <typename T>
+static bool check_cudnn_depthwise_workload_with_filter(const at::Tensor& input, T stride, const at::Tensor& weight) {
+  static int sm = at::detail::getCUDAHooks().getDeviceCapability();
```
Comment (drisspg) [inline aten/src/ATen/autoheuristic/DepthwiseConvHeuristic.h:12]: why do we need this instead of getting int form input device? is it faster?

update -> I guess its because we are in aten?

File: aten/src/ATen/autoheuristic/DepthwiseConvHeuristic.h [hunk 7]
```diff
@@ -0,0 +1,2727 @@
...
+  // only 1/3/5 filter
+  if (filter != 1 && filter != 3 && filter != 5) return false;
+  // we don't enforce square input but only check width to reduce heuristic space
+  if (at::symint::size<T>(input, 3) < 7) return false; // min width 7
+  auto w = at::symint::size<T>(input, 3);
+  auto ch = at::symint::size<T>(input, 1);
+  auto bs = at::symint::size<T>(input, 0);
+
+  // auto-generated heuristic decision tree, tolerance = 0.0
+  if (stride <= 1) {
```
Comment (drisspg) [inline aten/src/ATen/autoheuristic/DepthwiseConvHeuristic.h:34]: did we have any hold out set for testing ? like curious how well this tree is doing / other classification metrics

**Review by drisspg — COMMENTED** [review]
Did we test with symints as well?

**Review by drisspg — COMMENTED** [review]
I think it might be cleaner to have a tree per sm, although more jagged I feel a lil more confident in edge case perf behavior, wdyt?

**Comment by eqy** [comment]
@drisspg I agree modulo SMs that might be aliased according to cuDNN will check on this
